### PR TITLE
feat: add support for uploads from anonymous users

### PR DESCRIPTION
**Pull request by fiam**
- Add a new per-profile field, which defaults to requiring authentication
- Add end to end tests to nodejs-file-upload example, which also validates
that the upload hooks run.

### CODE REVIEW

**Review by Pagebakers — CHANGES_REQUESTED** [review]
I think the parameter name we use with functions is `requiresAuthentication` let's make sure we use the same (I couldn't find it in the docs btw, please verify)

I've checked the source code, that's the parameter name. I'll update it.

**Comment by fiam** [comment]
CI should pass once this is rebased on top of #631

**Comment by fiam** [comment]
> I think the parameter name we use with functions is `requiresAuthentication` let's make sure we use the same (I couldn't find it in the docs btw, please verify)
> 
> I've checked the source code, that's the parameter name. I'll update it.

Changed, PTAL

**Review by YuriBuerov — APPROVED** [review]

File: pkg/s3uploadclient/s3uploadclient.go [hunk 1]
```diff
@@ -50,6 +50,8 @@ type preparedProfile struct {
 // UploadProfile specifies options like maximum file size and allowed
 // extensions for a given upload profile
 type UploadProfile struct {
+	// Wether the profile requires authentication to upload a file
```
Comment (StarpTech) [inline pkg/s3uploadclient/s3uploadclient.go:53]: Typo

File: pkg/s3uploadclient/s3uploadclient.go [hunk 2]
```diff
@@ -408,6 +415,18 @@ func (s *S3UploadClient) handlePart(ctx context.Context, r *http.Request, part *
 	return info, nil
 }
 
+func (s *S3UploadClient) hasRequiredAuthentication(w http.ResponseWriter, r *http.Request) bool {
+	// Don't check for a valid profile name here. Since the default is requiring users
+	// to be authenticated, we can just check wether there's a profile. If the name
+	// doesn't exist we'll get an error later.
+	_, profile, _ := s.uploadProfile(r)
+	if (profile == nil || profile.RequiresAuthentication) && authentication.UserFromContext(r.Context()) == nil {
```
Comment (StarpTech) [inline pkg/s3uploadclient/s3uploadclient.go:423]: That contains many error cases. I prefer to handle each of them individually with some logs.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/s3uploadclient/s3uploadclient.go (hunk 1 above)
Comment (fiam) [reply pkg/s3uploadclient/s3uploadclient.go:53]: Thanks, fixed

File: pkg/s3uploadclient/s3uploadclient.go (hunk 2 above)
Comment (fiam) [reply pkg/s3uploadclient/s3uploadclient.go]: I've separated it into two checks and added a log message. PTAL

File: pkg/s3uploadclient/s3uploadclient.go [hunk 3]
```diff
@@ -408,6 +417,23 @@ func (s *S3UploadClient) handlePart(ctx context.Context, r *http.Request, part *
...
 
+func (s *S3UploadClient) hasRequiredAuthentication(w http.ResponseWriter, r *http.Request) bool {
+	// Don't check for a valid profile name here. Since the default is requiring users
+	// to be authenticated, we can just check if there's a profile. If the name
+	// doesn't exist we'll get an error later.
+	profileName, profile, _ := s.uploadProfile(r)
+	if profile == nil || profile.RequiresAuthentication {
+		if authentication.UserFromContext(r.Context()) == nil {
+			if s.logger != nil {
+				s.logger.Info("refusing upload from anonymous user", zap.String("provider", s.name), zap.String("profile", profileName))
```
Comment (StarpTech) [inline pkg/s3uploadclient/s3uploadclient.go:428]: I'd prefer debug.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/s3uploadclient/s3uploadclient.go (hunk 3 above)
Comment (fiam) [reply pkg/s3uploadclient/s3uploadclient.go]: Changed to Debug

**Review by Pagebakers — APPROVED** [review]

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**